File: maestro/scripts/maestro_fetch.py
#!/usr/bin/env python3
"""Update Maestro's known bundle index from a skill-bundle-folder."""

from __future__ import annotations

import argparse
import collections
import json
import pathlib
import re
import sys
from typing import Dict, Iterable, List, Tuple


START = " "
END = " "
ROOT_START = " "
ROOT_END = " "
KNOWN_HEADING = "## Known Bundles"
SELECTED_BUNDLE_PREFIX = "Selected bundle folder:"

TEXT_FILE_NAMES = {
    "skill.md",
    "readme.md",
    "readme.txt",
    "metadata.json",
    "agents/openai.yaml",
}
TEXT_SUFFIXES = {".md", ".markdown", ".skill", ".yaml", ".yml", ".txt", ".json"}
PATH_SIGNAL_SUFFIXES = TEXT_SUFFIXES | {".csv", ".tsv"}
SKIP_DIRS = {".git", ".hg", ".svn", "__pycache__", "node_modules", "dist", "build", ".next"}
MAX_DISCOVERY_FILES = 48
MAX_PATH_SIGNAL_FILES = 240
DOMAIN_SUFFIXES = (".com", ".org", ".net", ".io", ".dev", ".app")
CONFIG_FILE_NAME = "maestro_state.json"


def config_file(skill_file: pathlib.Path) -> pathlib.Path:
    return skill_file.resolve().parent / CONFIG_FILE_NAME


def load_saved_bundle_root(skill_file: pathlib.Path):
    try:
        data = json.loads(read_text(config_file(skill_file), limit=20000) or "{}")
    except json.JSONDecodeError:
        return None
    raw_path = data.get("bundle_root")
    if not raw_path:
        return None
    try:
        return pathlib.Path(str(raw_path)).expanduser().resolve()
    except OSError:
        return None


def save_bundle_root(skill_file: pathlib.Path, bundle_root: pathlib.Path) -> None:
    data = {"bundle_root": str(bundle_root.expanduser().resolve())}
    config_file(skill_file).write_text(json.dumps(data, indent=2) + "\n", encoding="utf-8", newline="\n")
    update_bundle_root_block(skill_file, bundle_root)

STOPWORDS = {
    "about",
    "added",
    "all",
    "and",
    "any",
    "are",
    "agent",
    "agents",
    "bundle",
    "bundles",
    "codex",
    "code",
    "community",
    "complete",
    "content",
    "core",
    "critical",
    "create",
    "csv",
    "data",
    "date",
    "default",
    "description",
    "directory",
    "end",
    "for",
    "file",
    "files",
    "folder",
    "folders",
    "from",
    "guide",
    "https",
    "guidelines",
    "help",
    "instruction",
    "instructions",
    "maestro",
    "metadata",
    "medium",
    "needed",
    "new",
    "name",
    "project",
    "reference",
    "references",
    "required",
    "requirements",
    "risk",
    "run",
    "pro",
    "readme",
    "resources",
    "scripts",
    "source",
    "skill",
    "skill.md",
    "skills",
    "task",
    "tasks",
    "that",
    "the",
    "this",
    "todo",
    "tool",
    "tools",
    "unknown",
    "use",
    "user",
    "version",
    "when",
    "work",
    "apply",
    "bash",
    "best",
    "development",
    "documentation",
    "github",
    "high",
    "how",
    "including",
    "install",
    "limitations",
    "not",
    "pattern",
    "patterns",
    "practice",
    "practices",
    "review",
    "script",
    "step",
    "steps",
    "you",
    "with",
}


def words(text: str) -> Iterable[str]:
    for token in re.findall(r"[A-Za-z][A-Za-z0-9+.#-]{1,}", text.lower()):
        token = token.strip("-_.")
        if len(token) < 3 or token in STOPWORDS:
            continue
        if token.endswith(DOMAIN_SUFFIXES):
            continue
        yield token


def read_text(path: pathlib.Path, limit: int = 24000) -> str:
    try:
        data = path.read_bytes()[:limit]
    except OSError:
        return ""
    for encoding in ("utf-8", "utf-16", "latin-1"):
        try:
            return data.decode(encoding, errors="ignore")
        except UnicodeError:
            continue
    return ""


def is_skipped(path: pathlib.Path) -> bool:
    return any(part.lower() in SKIP_DIRS or part.startswith(".") for part in path.parts)


def add_candidate(candidates: List[pathlib.Path], path: pathlib.Path) -> None:
    if path.exists() and path.is_file() and path not in candidates:
        candidates.append(path)


def scope_candidates(scope: pathlib.Path) -> List[pathlib.Path]:
    candidates: List[pathlib.Path] = []
    for rel in ("SKILL.md", "README.md", "README.txt", "metadata.json", "agents/openai.yaml"):
        add_candidate(candidates, scope / rel)
    try:
        children = sorted(child for child in scope.iterdir() if child.is_file())
    except OSError:
        return candidates
    candidates.extend(children)
    return candidates

def discovery_files(bundle: pathlib.Path) -> List[pathlib.Path]:
    found: List[pathlib.Path] = []
    seen = set()
    candidates = scope_candidates(bundle)
    try:
        subdirs = sorted(child for child in bundle.iterdir() if child.is_dir() and not is_skipped(child.relative_to(bundle)))
    except OSError:
        subdirs = []
    for subdir in subdirs:
        candidates.extend(scope_candidates(subdir))

    for child in candidates:
        rel = child.relative_to(bundle).as_posix().lower() if child.exists() else ""
        if not child.is_file() or child in seen:
            continue
        if rel in TEXT_FILE_NAMES or child.suffix.lower() in TEXT_SUFFIXES:
            found.append(child)
            seen.add(child)
        if len(found) >= MAX_DISCOVERY_FILES:
            break
    return found


def path_signal_files(bundle: pathlib.Path) -> Iterable[pathlib.Path]:
    count = 0
    for path in sorted(bundle.rglob("*")):
        try:
            rel = path.relative_to(bundle)
        except ValueError:
            continue
        if is_skipped(rel) or path.is_dir() or path.suffix.lower() not in PATH_SIGNAL_SUFFIXES:
            continue
        yield path
        count += 1
        if count >= MAX_PATH_SIGNAL_FILES:
            break


def score_path(score: collections.Counter[str], rel: pathlib.PurePath, weight: int) -> None:
    for part in rel.parts:
        stem = pathlib.PurePath(part).stem
        for token in words(stem.replace("-", " ").replace("_", " ")):
            score[token] += weight


def infer_keywords(bundle: pathlib.Path, max_keywords: int = 32) -> List[str]:
    score: collections.Counter[str] = collections.Counter()

    score[bundle.name.lower()] += 14
    name_text = bundle.name.replace("-", " ").replace("_", " ")
    for token in words(name_text):
        score[token] += 8
    for part in re.split(r"[-_\s]+", bundle.name.lower()):
        if part and len(part) >= 3 and part not in STOPWORDS:
            score[part] += 10

    for child in sorted(bundle.iterdir()):
        weight = 6 if child.is_dir() else 3
        for token in words(child.stem.replace("-", " ").replace("_", " ")):
            score[token] += weight

    for path in path_signal_files(bundle):
        score_path(score, path.relative_to(bundle), 3)

    for path in discovery_files(bundle):
        rel = path.relative_to(bundle).as_posix()
        for token in words(rel.replace("/", " ").replace("-", " ").replace("_", " ")):
            score[token] += 2
        text = read_text(path)
        frontmatter = re.search(r"^---\s*(.*?)\s*---", text, re.S | re.M)
        if frontmatter:
            for token in words(frontmatter.group(1)):
                score[token] += 5
        for heading in re.findall(r"^#{1,3}\s+(.+)$", text, re.M):
            for token in words(heading):
                score[token] += 4
        for command in re.findall(r"`(/?[A-Za-z0-9_.:-]+)`", text):
            for token in words(command):
                score[token] += 4
        for token in words(text[:6000]):
            score[token] += 1

    ranked = sorted(score.items(), key=lambda item: (-item[1], item[0]))
    return [token for token, _ in ranked[:max_keywords]]


def parse_existing(block: str) -> Dict[str, Tuple[List[str], str]]:
    rows: Dict[str, Tuple[List[str], str]] = {}
    for line in block.splitlines():
        line = line.strip()
        if not line.startswith("|") or "---" in line or "Bundle" in line:
            continue
        cells = [cell.strip() for cell in line.strip("|").split("|")]
        if len(cells) < 3:
            continue
        name, keyword_text, path = cells[:3]
        if not name:
            continue
        keywords = []
        for item in keyword_text.split(","):
            item = item.strip()
            if item and item.lower() not in STOPWORDS and not item.lower().endswith(DOMAIN_SUFFIXES):
                keywords.append(item)
        rows[name] = (keywords, path.strip("`"))
    return rows


def render_bundle_root(bundle_root: pathlib.Path = None) -> str:
    if bundle_root is None:
        return "Selected bundle folder: Not configured. Run `/maestro-switch <skill-bundle-folder>`."
    return f"Selected bundle folder: `{str(bundle_root.expanduser().resolve())}`"


def known_bundles_bounds(text: str) -> Tuple[int, int, int]:
    heading = re.search(r"(?m)^## Known Bundles\s*$", text)
    if not heading:
        raise SystemExit("Known Bundles section missing")
    content_start = heading.end()
    next_heading = re.search(r"(?m)^##\s+.+$", text[content_start:])
    content_end = content_start + next_heading.start() if next_heading else len(text)
    return heading.start(), content_start, content_end


def known_bundles_block(text: str) -> str:
    _, content_start, content_end = known_bundles_bounds(text)
    return text[content_start:content_end]


def render_known_bundles_section(rows: Dict[str, Tuple[List[str], str]], bundle_root: pathlib.Path = None) -> str:
    return KNOWN_HEADING + "\n" + render_bundle_root(bundle_root) + "\n\n" + render_index(rows) + "\n\n"


def set_known_bundles_section(text: str, rows: Dict[str, Tuple[List[str], str]], bundle_root: pathlib.Path = None) -> str:
    section_start, _, content_end = known_bundles_bounds(text)
    return text[:section_start] + render_known_bundles_section(rows, bundle_root) + text[content_end:].lstrip("\r\n")


def set_bundle_root_block(text: str, bundle_root: pathlib.Path = None) -> str:
    rows = parse_existing(known_bundles_block(text))
    return set_known_bundles_section(text, rows, bundle_root)


def update_bundle_root_block(skill_file: pathlib.Path, bundle_root: pathlib.Path = None) -> None:
    text = read_text(skill_file, limit=1_000_000)
    if not text:
        return
    skill_file.write_text(set_bundle_root_block(text, bundle_root), encoding="utf-8", newline="\n")


def render_index(rows: Dict[str, Tuple[List[str], str]]) -> str:
    if not rows:
        return "No bundles indexed yet. Run `/maestro-fetch <skill-bundle-folder>`."
    lines = [
        "| Bundle | Keywords | Path |",
        "| --- | --- | --- |",
    ]
    for name in sorted(rows, key=str.lower):
        keywords, path = rows[name]
        clean_keywords = ", ".join(dict.fromkeys(keywords))
        escaped_path = path.replace("|", r"\|")
        lines.append(f"| {name} | {clean_keywords} | `{escaped_path}` |")
    return "\n".join(lines)


def update_skill(skill_file: pathlib.Path, bundle_root: pathlib.Path, keep_missing: bool = False) -> Tuple[int, int, int, int]:
    text = read_text(skill_file, limit=1_000_000)
    old_block = known_bundles_block(text)
    rows = parse_existing(old_block)
    before_count = len(rows)
    existing_names = set(rows)
    scanned_rows: Dict[str, Tuple[List[str], str]] = {}

    for bundle in sorted(bundle_root.iterdir()):
        if not bundle.is_dir() or bundle.name.startswith("."):
            continue
        inferred = infer_keywords(bundle)
        existing_keywords, _ = rows.get(bundle.name, ([], ""))
        merged_keywords = list(dict.fromkeys(existing_keywords + inferred))[:36]
        scanned_rows[bundle.name] = (merged_keywords, str(bundle.resolve()))

    scanned_names = set(scanned_rows)
    added_count = len(scanned_names - existing_names)
    pruned_count = 0 if keep_missing else len(existing_names - scanned_names)
    if keep_missing:
        rows.update(scanned_rows)
        next_rows = rows
    else:
        next_rows = scanned_rows

    new_text = set_known_bundles_section(text, next_rows, bundle_root)
    skill_file.write_text(new_text, encoding="utf-8", newline="\n")
    return before_count, len(next_rows), added_count, pruned_count


def main() -> int:
    parser = argparse.ArgumentParser(description="Update Maestro bundle index.")
    parser.add_argument("bundle_root", nargs="?", help="Folder containing immediate child skill bundles. If omitted, uses saved Maestro bundle root.")
    parser.add_argument(
        "--skill-file",
        default=str(pathlib.Path(__file__).resolve().parents[1] / "SKILL.md"),
        help="Maestro SKILL.md path. Defaults to this skill's SKILL.md.",
    )
    parser.add_argument(
        "--keep-missing",
        action="store_true",
        help="Keep existing index rows whose bundle folders are missing. Default prunes them.",
    )
    parser.add_argument(
        "--no-save-root",
        action="store_true",
        help="Do not save this bundle root as Maestro's default skill-bundle-folder.",
    )
    args = parser.parse_args()

    skill_file = pathlib.Path(args.skill_file).expanduser().resolve()
    if not skill_file.is_file():
        raise SystemExit(f"skill file not readable: {skill_file}")

    if args.bundle_root:
        bundle_root = pathlib.Path(args.bundle_root).expanduser().resolve()
    else:
        bundle_root = load_saved_bundle_root(skill_file)
        if bundle_root is None:
            raise SystemExit("bundle root not configured. Run /maestro-switch <skill-bundle-folder> or pass bundle_root.")

    if not bundle_root.is_dir():
        raise SystemExit(f"bundle root not readable: {bundle_root}")

    if not args.no_save_root:
        save_bundle_root(skill_file, bundle_root)

    _, after, added, pruned = update_skill(skill_file, bundle_root, keep_missing=args.keep_missing)
    print(f"indexed {after} bundles ({added} new, {pruned} pruned) in {skill_file} using {bundle_root}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: maestro/scripts/maestro_fetch.js
#!/usr/bin/env node
"use strict";
/**
 * Update Maestro's known bundle index from a skill-bundle-folder.
 * Maestro bundle index refresh script. Zero external dependencies.
 */

const fs   = require("fs");
const path = require("path");

// ── Markers & constants ──────────────────────────────────────────────
const KNOWN_HEADING          = "## Known Bundles";
const SELECTED_BUNDLE_PREFIX = "Selected bundle folder:";
const CONFIG_FILE_NAME       = "maestro_state.json";

const TEXT_FILE_NAMES = new Set([
  "skill.md", "readme.md", "readme.txt", "metadata.json", "agents/openai.yaml",
]);
const TEXT_SUFFIXES         = new Set([".md", ".markdown", ".skill", ".yaml", ".yml", ".txt", ".json"]);
const PATH_SIGNAL_SUFFIXES = new Set([...TEXT_SUFFIXES, ".csv", ".tsv"]);
const SKIP_DIRS            = new Set([".git", ".hg", ".svn", "__pycache__", "node_modules", "dist", "build", ".next"]);
const MAX_DISCOVERY_FILES     = 48;
const MAX_PATH_SIGNAL_FILES   = 240;
const MIN_KEYWORDS            = 10;
const MAX_MERGED_KEYWORDS     = 36;
const DOMAIN_SUFFIXES         = [".com", ".org", ".net", ".io", ".dev", ".app"];

const STOPWORDS = new Set([
  "about","added","all","and","any","are","agent","agents","bundle","bundles",
  "codex","code","community","complete","content","core","critical","create",
  "csv","data","date","default","description","directory","end","for","file",
  "files","folder","folders","from","guide","https","guidelines","help",
  "instruction","instructions","maestro","metadata","medium","needed","new",
  "name","project","reference","references","required","requirements","risk",
  "run","pro","readme","resources","scripts","source","skill","skill.md",
  "skills","task","tasks","that","the","this","todo","tool","tools","unknown",
  "use","user","version","when","work","apply","bash","best","development",
  "documentation","github","high","how","including","install","limitations",
  "not","pattern","patterns","practice","practices","review","script","step",
  "steps","you","with",
]);

// ── Helpers ───────────────────────────────────────────────────────────

function configFile(skillFile) {
  return path.join(path.dirname(path.resolve(skillFile)), CONFIG_FILE_NAME);
}

function readText(filePath, limit = 24000) {
  try {
    const buf = Buffer.alloc(limit);
    const fd  = fs.openSync(filePath, "r");
    const n   = fs.readSync(fd, buf, 0, limit, 0);
    fs.closeSync(fd);
    return buf.slice(0, n).toString("utf-8");
  } catch {
    return "";
  }
}

function words(text) {
  const matches = text.toLowerCase().match(/[a-z][a-z0-9+.#-]{1,}/g) || [];
  const result = [];
  for (let token of matches) {
    token = token.replace(/^[-_.]+|[-_.]+$/g, "");
    if (token.length < 3 || STOPWORDS.has(token)) continue;
    if (DOMAIN_SUFFIXES.some(s => token.endsWith(s))) continue;
    result.push(token);
  }
  return result;
}

function cleanKeyword(keyword) {
  const cleaned = String(keyword || "")
    .toLowerCase()
    .replace(/^[-_.\s]+|[-_.\s]+$/g, "")
    .replace(/\s+/g, "-");
  if (cleaned.length < 3 || STOPWORDS.has(cleaned)) return "";
  if (DOMAIN_SUFFIXES.some(s => cleaned.endsWith(s))) return "";
  return cleaned;
}

function addScore(score, keyword, weight) {
  const cleaned = cleanKeyword(keyword);
  if (!cleaned) return;
  score[cleaned] = (score[cleaned] || 0) + weight;
}

function scorePhrase(score, text, weight) {
  const normalized = String(text || "")
    .replace(/([a-z])([A-Z])/g, "$1 $2")
    .replace(/[^A-Za-z0-9+.#_-]+/g, " ")
    .trim();
  if (!normalized) return;

  const compact = cleanKeyword(normalized.replace(/\s+/g, "-"));
  if (compact && compact.includes("-")) addScore(score, compact, weight);

  for (const token of words(normalized.replace(/[-_]/g, " "))) {
    addScore(score, token, weight);
  }
}

function isSkipped(relPath) {
  const parts = relPath.split(path.sep);
  return parts.some(p => SKIP_DIRS.has(p.toLowerCase()) || p.startsWith("."));
}

// ── Config persistence ───────────────────────────────────────────────

function loadSavedBundleRoot(skillFile) {
  const cfgPath = configFile(skillFile);
  const text = readText(cfgPath, 20000);
  if (!text) return null;
  try {
    const data = JSON.parse(text);
    if (!data.bundle_root) return null;
    return path.resolve(data.bundle_root);
  } catch {
    return null;
  }
}

function saveBundleRoot(skillFile, bundleRoot) {
  const resolved = path.resolve(bundleRoot);
  const data = { bundle_root: resolved };
  fs.writeFileSync(configFile(skillFile), JSON.stringify(data, null, 2) + "\n", "utf-8");
  updateBundleRootBlock(skillFile, resolved);
}

// ── File discovery ───────────────────────────────────────────────────

function addCandidate(candidates, filePath) {
  try {
    if (fs.existsSync(filePath) && fs.statSync(filePath).isFile() && !candidates.includes(filePath)) {
      candidates.push(filePath);
    }
  } catch { /* skip */ }
}

function scopeCandidates(scope) {
  const candidates = [];
  for (const rel of ["SKILL.md", "README.md", "README.txt", "metadata.json", path.join("agents", "openai.yaml")]) {
    addCandidate(candidates, path.join(scope, rel));
  }
  try {
    const children = fs.readdirSync(scope)
      .map(c => path.join(scope, c))
      .filter(c => { try { return fs.statSync(c).isFile(); } catch { return false; } })
      .sort();
    candidates.push(...children);
  } catch { /* skip */ }
  return candidates;
}

function discoveryFiles(bundle) {
  const found = [];
  const seen  = new Set();
  const candidates = scopeCandidates(bundle);

  let subdirs = [];
  try {
    subdirs = fs.readdirSync(bundle)
      .map(c => path.join(bundle, c))
      .filter(c => {
        try {
          if (!fs.statSync(c).isDirectory()) return false;
          const rel = path.relative(bundle, c);
          return !isSkipped(rel);
        } catch { return false; }
      })
      .sort();
  } catch { /* skip */ }

  for (const subdir of subdirs) {
    candidates.push(...scopeCandidates(subdir));
  }

  for (const child of candidates) {
    if (seen.has(child)) continue;
    try { if (!fs.statSync(child).isFile()) continue; } catch { continue; }
    const rel = path.relative(bundle, child).replace(/\\/g, "/").toLowerCase();
    if (TEXT_FILE_NAMES.has(rel) || TEXT_SUFFIXES.has(path.extname(child).toLowerCase())) {
      found.push(child);
      seen.add(child);
    }
    if (found.length >= MAX_DISCOVERY_FILES) break;
  }
  return found;
}

function* pathSignalFiles(bundle) {
  let count = 0;
  const walk = (dir) => {
    let entries;
    try { entries = fs.readdirSync(dir, { withFileTypes: true }).sort((a,b) => a.name.localeCompare(b.name)); } catch { return []; }
    const results = [];
    for (const entry of entries) {
      const full = path.join(dir, entry.name);
      const rel  = path.relative(bundle, full);
      if (entry.isDirectory()) {
        if (!isSkipped(rel)) results.push(...walk(full));
      } else if (entry.isFile() && PATH_SIGNAL_SUFFIXES.has(path.extname(entry.name).toLowerCase())) {
        results.push(full);
      }
    }
    return results;
  };
  for (const p of walk(bundle)) {
    yield p;
    count++;
    if (count >= MAX_PATH_SIGNAL_FILES) return;
  }
}

// ── Keyword inference ────────────────────────────────────────────────

function scorePath(score, relPath, weight) {
  const parts = relPath.split(path.sep);
  for (const part of parts) {
    const stem = path.parse(part).name;
    scorePhrase(score, stem, weight);
  }
}

function inferKeywords(bundle, maxKeywords = 32) {
  const score = {};

  // Bundle folder name
  const bundleName = path.basename(bundle).toLowerCase();
  addScore(score, bundleName, 14);
  scorePhrase(score, path.basename(bundle), 12);
  for (const part of path.basename(bundle).toLowerCase().split(/[-_\s]+/)) {
    if (part && part.length >= 3 && !STOPWORDS.has(part)) {
      addScore(score, part, 10);
    }
  }

  // Direct children
  try {
    const children = fs.readdirSync(bundle).sort();
    for (const child of children) {
      const full = path.join(bundle, child);
      let weight;
      try { weight = fs.statSync(full).isDirectory() ? 6 : 3; } catch { weight = 3; }
      const stem = path.parse(child).name;
      scorePhrase(score, stem, weight);
    }
  } catch { /* skip */ }

  // Path signal files
  for (const p of pathSignalFiles(bundle)) {
    scorePath(score, path.relative(bundle, p), 3);
  }

  // Discovery files content
  for (const p of discoveryFiles(bundle)) {
    const rel = path.relative(bundle, p).replace(/\\/g, "/");
    scorePath(score, rel, 2);
    const text = readText(p);
    const fmMatch = text.match(/^---\s*([\s\S]*?)\s*---/m);
    if (fmMatch) {
      for (const token of words(fmMatch[1])) {
        addScore(score, token, 5);
      }
    }
    const headings = text.match(/^#{1,3}\s+(.+)$/gm) || [];
    for (const heading of headings) {
      scorePhrase(score, heading.replace(/^#{1,3}\s+/, ""), 4);
    }
    const commands = text.match(/`(\/?[A-Za-z0-9_.:-]+)`/g) || [];
    for (const cmd of commands) {
      scorePhrase(score, cmd.replace(/`/g, ""), 4);
    }
    for (const token of words(text.slice(0, 6000))) {
      addScore(score, token, 1);
    }
  }

  const ranked = Object.entries(score).sort((a, b) => b[1] - a[1] || a[0].localeCompare(b[0]));
  return ranked.slice(0, maxKeywords).map(([token]) => token);
}

function normalizeKeywords(keywords) {
  const result = [];
  const seen = new Set();
  for (const keyword of keywords || []) {
    const cleaned = cleanKeyword(keyword);
    if (!cleaned || seen.has(cleaned)) continue;
    seen.add(cleaned);
    result.push(cleaned);
  }
  return result;
}

function padKeywords(keywords, bundleName, minKeywords = MIN_KEYWORDS) {
  const padded = normalizeKeywords(keywords);
  const seen = new Set(padded);
  const bases = normalizeKeywords([bundleName, ...String(bundleName || "").split(/[-_\s]+/)]);
  const base = bases[0] || "bundle";
  const suffixes = [
    "workflow", "patterns", "reference", "tools", "automation",
    "configuration", "commands", "routing", "docs", "usage", "setup", "examples",
  ];

  for (const suffix of suffixes) {
    if (padded.length >= minKeywords) break;
    const keyword = cleanKeyword(`${base}-${suffix}`);
    if (!keyword || seen.has(keyword)) continue;
    seen.add(keyword);
    padded.push(keyword);
  }

  return padded;
}

function mergeKeywords(existingKeywords, inferredKeywords, bundleName = "", minKeywords = MIN_KEYWORDS, maxKeywords = MAX_MERGED_KEYWORDS) {
  const merged = normalizeKeywords(existingKeywords);
  const seen = new Set(merged);

  for (const keyword of normalizeKeywords(inferredKeywords)) {
    if (seen.has(keyword)) continue;
    seen.add(keyword);
    merged.push(keyword);
  }

  return padKeywords(merged, bundleName, minKeywords).slice(0, Math.max(minKeywords, maxKeywords));
}

// ── Known Bundles section parsing & rendering ────────────────────────

function parseExisting(block) {
  const rows = {};
  for (const line of block.split("\n")) {
    const trimmed = line.trim();
    if (!trimmed.startsWith("|") || trimmed.includes("---") || /^\|\s*Bundle\s*\|/.test(trimmed)) continue;
    const cells = trimmed.replace(/^\||\|$/g, "").split("|").map(c => c.trim());
    if (cells.length < 3) continue;
    const [name, keywordText, rawPath] = cells;
    if (!name) continue;
    const keywords = normalizeKeywords(keywordText.split(","));
    rows[name] = [keywords, rawPath.replace(/`/g, "")];
  }
  return rows;
}

function renderBundleRoot(bundleRoot) {
  if (!bundleRoot) {
    return "Selected bundle folder: Not configured. Run `/maestro-set <skill-bundle-folder>`.";
  }
  return `Selected bundle folder: \`${path.resolve(bundleRoot)}\``;
}

function knownBundlesBounds(text) {
  const headingMatch = text.match(/^## Known Bundles\s*$/m);
  if (!headingMatch) {
    throw new Error("Known Bundles section missing");
  }
  const headingStart  = headingMatch.index;
  const contentStart  = headingStart + headingMatch[0].length;
  const rest          = text.slice(contentStart);
  const nextHeading   = rest.match(/^##\s+.+$/m);
  const contentEnd    = nextHeading ? contentStart + nextHeading.index : text.length;
  return [headingStart, contentStart, contentEnd];
}

function knownBundlesBlock(text) {
  const [, contentStart, contentEnd] = knownBundlesBounds(text);
  return text.slice(contentStart, contentEnd);
}

function renderIndex(rows) {
  const names = Object.keys(rows).sort((a, b) => a.toLowerCase().localeCompare(b.toLowerCase()));
  if (names.length === 0) {
    return "No bundles indexed yet. Run `/maestro-fetch <skill-bundle-folder>`.";
  }
  const lines = [
    "| Bundle | Keywords | Path |",
    "| --- | --- | --- |",
  ];
  for (const name of names) {
    const [keywords, p] = rows[name];
    const cleanKeywords = normalizeKeywords(keywords).join(", ");
    const escapedPath = p.replace(/\|/g, "\\|");
    lines.push(`| ${name} | ${cleanKeywords} | \`${escapedPath}\` |`);
  }
  return lines.join("\n");
}

function renderKnownBundlesSection(rows, bundleRoot) {
  return KNOWN_HEADING + "\n" + renderBundleRoot(bundleRoot) + "\n\n" + renderIndex(rows) + "\n\n";
}

function setKnownBundlesSection(text, rows, bundleRoot) {
  const [sectionStart, , contentEnd] = knownBundlesBounds(text);
  return text.slice(0, sectionStart) + renderKnownBundlesSection(rows, bundleRoot) + text.slice(contentEnd).replace(/^[\r\n]+/, "");
}

function setBundleRootBlock(text, bundleRoot) {
  const rows = parseExisting(knownBundlesBlock(text));
  return setKnownBundlesSection(text, rows, bundleRoot);
}

function updateBundleRootBlock(skillFile, bundleRoot) {
  const text = readText(skillFile, 1000000);
  if (!text) return;
  fs.writeFileSync(skillFile, setBundleRootBlock(text, bundleRoot), "utf-8");
}

// ── Main update logic ────────────────────────────────────────────────

function updateSkill(skillFile, bundleRoot, keepMissing = false) {
  const text     = readText(skillFile, 1000000);
  const oldBlock = knownBundlesBlock(text);
  const rows     = parseExisting(oldBlock);
  const beforeCount   = Object.keys(rows).length;
  const existingNames = new Set(Object.keys(rows));
  const scannedRows   = {};

  let entries;
  try { entries = fs.readdirSync(bundleRoot).sort(); } catch { entries = []; }
  for (const entry of entries) {
    const bundlePath = path.join(bundleRoot, entry);
    try { if (!fs.statSync(bundlePath).isDirectory() || entry.startsWith(".")) continue; } catch { continue; }
    const inferred = inferKeywords(bundlePath);
    const [existingKeywords] = rows[entry] || [[], ""];
    const mergedKeywords = mergeKeywords(existingKeywords, inferred, entry);
    scannedRows[entry] = [mergedKeywords, path.resolve(bundlePath)];
  }

  const scannedNames = new Set(Object.keys(scannedRows));
  const addedCount  = [...scannedNames].filter(n => !existingNames.has(n)).length;
  const prunedCount = keepMissing ? 0 : [...existingNames].filter(n => !scannedNames.has(n)).length;

  let nextRows;
  if (keepMissing) {
    nextRows = { ...rows, ...scannedRows };
  } else {
    nextRows = scannedRows;
  }

  const newText = setKnownBundlesSection(text, nextRows, bundleRoot);
  fs.writeFileSync(skillFile, newText, "utf-8");
  return [beforeCount, Object.keys(nextRows).length, addedCount, prunedCount];
}

// ── CLI ──────────────────────────────────────────────────────────────

function main() {
  const args = process.argv.slice(2);
  let bundleRootArg = null;
  let skillFileArg  = path.join(__dirname, "..", "SKILL.md");
  let keepMissing   = false;
  let noSaveRoot    = false;

  for (let i = 0; i < args.length; i++) {
    if (args[i] === "--skill-file" && args[i + 1]) { skillFileArg = args[++i]; }
    else if (args[i] === "--keep-missing") { keepMissing = true; }
    else if (args[i] === "--no-save-root") { noSaveRoot = true; }
    else if (!args[i].startsWith("-")) { bundleRootArg = args[i]; }
  }

  const skillFile = path.resolve(skillFileArg);
  if (!fs.existsSync(skillFile)) {
    console.error(`skill file not readable: ${skillFile}`);
    process.exit(1);
  }

  let bundleRoot;
  if (bundleRootArg) {
    bundleRoot = path.resolve(bundleRootArg);
  } else {
    bundleRoot = loadSavedBundleRoot(skillFile);
    if (!bundleRoot) {
      console.error("bundle root not configured. Run /maestro-set <skill-bundle-folder> or pass bundle_root.");
      process.exit(1);
    }
  }

  try {
    if (!fs.statSync(bundleRoot).isDirectory()) throw new Error();
  } catch {
    console.error(`bundle root not readable: ${bundleRoot}`);
    process.exit(1);
  }

  if (!noSaveRoot) {
    saveBundleRoot(skillFile, bundleRoot);
  }

  const [, after, added, pruned] = updateSkill(skillFile, bundleRoot, keepMissing);
  console.log(`indexed ${after} bundles (${added} new, ${pruned} pruned) in ${skillFile} using ${bundleRoot}`);
}



// ── Exports (for use by other scripts) ───────────────────────────────
module.exports = {
  configFile, readText, words, isSkipped, loadSavedBundleRoot, saveBundleRoot,
  discoveryFiles, pathSignalFiles, inferKeywords, parseExisting, renderBundleRoot,
  knownBundlesBounds, knownBundlesBlock, renderIndex, renderKnownBundlesSection,
  setKnownBundlesSection, setBundleRootBlock, updateBundleRootBlock, updateSkill,
  normalizeKeywords, padKeywords, mergeKeywords, STOPWORDS, DOMAIN_SUFFIXES, MIN_KEYWORDS,
};

if (require.main === module) {
  main();
}
